nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py

```diff
@@ -209,6 +230,143 @@ def debug(self, msg):
+            else:
+                self.get_logger().error('Failed to call map service')
+        else:
+            self.get_logger().error('Map service not available')
+
+    def get_laser_pose(self):
+        try:
+            if self.initial_pose is None:
+                return 0.0, 0.0, 0.0
+
+            if self.map is None:
+                self.get_map()
+                return 0.0, 0.0, 0.0
+
+            # Wait for transform and lookup
+            now = rclpy.time.Time()
+            transform = self.tf_buffer.lookup_transform(
+                'map', self.scan_frame_id, now, rclpy.duration.Duration(seconds=1.0))
+
+            # Extract pose information
+            x = transform.transform.translation.x
+            y = transform.transform.translation.y
+            rotation = transform.transform.rotation
+            theta = tf_transformations.euler_from_quaternion([
+                rotation.x,
+                rotation.y,
+                rotation.z,
+                rotation.w
+            ])[2]
+
+            return x, y, theta
+
+        except Exception as ex:
+            self.get_logger().error('Transform lookup failed: %s' % str(ex))
+            return 0.0, 0.0, 0.0
+
+    def get_laserscan(self, num_samples, x, y, theta):
+        for i in range(int(num_samples)):
+            if self.map is None:
+                self.scan_msg.ranges = [self.scan_msg.range_max] * num_samples
+                break
+            else:
+                curr_angle = theta + self.scan_msg.angle_min + i * self.scan_msg.angle_increment
+                self.scan_msg.ranges[i] = self.find_map_range(x, y, curr_angle)
+
+    def find_map_range(self, x, y, theta):
+        # Using "A Faster Voxel Traversal Algorithm for Ray Tracing" by Amanatides & Woo
+        # ======== Initialization Phase ========
+        origin = [x, y]  # u
+        direction = [math.cos(theta), math.sin(theta)]  # v
+
+        start_x, start_y = get_world2map_coordinates(x, y, self.map)
+
+        if (start_x < 0 or start_y < 0 or
+                start_x >= self.map.info.width or start_y >= self.map.info.height):
+            # Outside the map, find entry point
+            delta_x = abs(self.map.info.origin.position.x - x)
+            delta_y = abs(self.map.info.origin.position.y - y)
+            intersect_x = x + (direction[0] * delta_x)
+            intersect_y = y + (direction[1] * delta_y)
+            start_x, start_y = get_world2map_coordinates(intersect_x, intersect_y, self.map)
+
+        current = [start_x, start_y]  # X, Y
+        step = [0, 0]  # stepX, stepY
+        tMax = [0.0, 0.0]  # tMaxX, tMaxY
+        tDelta = [0.0, 0.0]  # tDeltaX, tDeltaY
+
+        voxel_border = [0.0, 0.0]
+        voxel_border[0], voxel_border[1] = get_map2world_coordinates(
+            current[0], current[1], self.map)
+        voxel_border[0] -= 0.5 * self.map.info.resolution  # This is the lower left voxel corner
```

Comment:
I'm not really grokking the voxel_border stuff from L313 to L337. Can you explain? I think the worldToMap and mapToWorld conversions should be handling the `0.5`'s fine itself. I don't get the other stuff

Reply:
I have tried to keep it same as original [reference](https://github.com/nobleo/mobile_robot_simulator/blob/master/src/laser_simulator.cpp)